cmd/rendmp/src/blackbox.rs

```diff
@@ -0,0 +1,774 @@
+                (6, "off"),
+                (5, "VOUT_OV_FAULT"),
+                (4, "IOUT_OC_FAULT"),
+                (3, "VIN_UV_FAULT"),
+                (2, "TEMPERATURE"),
+                (1, "CML"),
+                (0, "none of the above"),
+            ],
+            f,
+        )
+    }
+}
+
+#[derive(Debug, FromBytes)]
+struct StatusCml(u8);
+impl Display for StatusCml {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        format_faults_desc(
+            self.0,
+            &[
+                (15, "VOUT", "An output voltage fault or warning has occurred."),
+                (14, "IOUT", "An output current fault has occurred."),
+                (13, "INPUT", "An input voltage fault or warning has occurred."),
+                (12, "MFR_SPECIFIC", "A manufacturer specific fault or warning has occurred."),
+                (11, "POWER_GOOD#", "The POWER_GOOD signal is negated."),
+                (8, "Unknown", "A fault other than those described in bits [15:9] has occurred."),
+                (7, "Busy", "Device busy and unable to respond."),
+                (6, "OFF", "This bit is asserted if the unit is not providing power to the output, regardless of the reason, including simply not being enabled"),
+                (5, "VOUT_OV_FAULT", "An output overvoltage fault has occurred."),
+                (4, "IOUT_OC_FAULT", "An output overcurrent fault has occurred."),
+                (3, "VIN_UV_FAULT", "An input undervoltage fault has occurred."),
+                (2, "TEMPERATURE", "A temperature fault or warning has occurred."),
+                (1, "CML", "A communications, memory, or logic fault has occurred."),
+                (0, "None of the Above", "A status change other than those listed above has occurred."),
+            ],
+            f,
+        )
+    }
+}
+
+#[derive(Debug, FromBytes)]
+struct StatusMfr(u8);
+impl Display for StatusMfr {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        format_faults_desc(
+            self.0,
+            &[
+                (
+                    7,
+                    "ADCUNLOCK",
+                    "Invalid or unsupported PMBus Command was received.",
+                ),
+                (
+                    6,
+                    "PSYS, IIN Sense",
+                    "A PSYS and/or sensed IIN OC warning has occurred.",
+                ),
+                (5, "CFP Fault", "A CFP fault has occurred."),
+                (
+                    4,
+                    "Internal Temperature Fault",
+                    "The controller internal temp has exceeded 130°C.",
+                ),
+                (3, "BBEVENT", "A Black Box event occurred."),
+                (2, "LMSEVENT", "A Last Man Standing event has occurred."),
+                (
+                    1,
+                    "SPSFault",
+                    "An SPS overcurrent and/or over-temperature \
+                     event has occurred.",
```

Comment:
I think this will put a bunch of spaces in the string. Maybe `concat!()` instead to split across lines?

Reply:
Ha, yes, this is something I have done/do a lot of in Humility. ;)